compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -760,6 +766,13 @@ impl<'a, 'b, 'tcx> TypeVerifier<'a, 'b, 'tcx> {
         self.tcx().ty_error()
     }
 
+    fn get_ambient_variance(&self, context: PlaceContext) -> ty::Variance {
+        match context {
+            PlaceContext::MutatingUse(_) => ty::Invariant,
+            PlaceContext::NonMutatingUse(_) | PlaceContext::NonUse(_) => ty::Covariant,
```

Comment:
there's `NonUseContext::AscribeUserTy` which I feel should maybe be invariant :thinking:

I can also imagene us adding `NonMutatingUse` in the future which should be invariant. Can you also exhaustively match on the `NonMutatingUseContext` and probably make all `NonUse` invariant. Don't know how to write an example where `Covariance` is needed here.

Considering that covariance can be unsound if used incorrectly, we should be as defensive as possible here.

Reply:
Sorry for the late answer. 

We would need `NonUseContext::AscribeUserTy` for field projections on enums in match statements where we always insert user type projections I believe, e.g. 

```
struct Inv<'a>(&'a mut &'a ());
enum Foo<T> {
    Bar,
    Var(T),
}
type Supertype = Foo<for<'a> fn(Inv<'a>, Inv<'a>)>;

fn foo_nested(x: Foo<Foo<for<'a, 'b> fn(Inv<'a>, Inv<'b>)>>) {
    match x {
        Foo::Bar => {}
        Foo::Var(Supertype::Bar) => {}
        Foo::Var(Supertype::Var(x)) => {}
    }
}
```

would not compile if we would use `ty::Invariant` on `NonUse(AscribeUserTy)`.

I agree that we should be very conservative here, but I also can't see how using covariance here might be unsound. Can you show that it is? I've kept this as covariant for now.

Reply:
my worry is that we use `AscribeUserTy` in a position which isn't covariant. E.g. only for arguments of a function we're calling. Thinking more about it I actually can't imagine that happening :thinking:

:+1:

Reply:
The `AscribeUserType` statement contains a variance. Should it be carried by `NonUseContext::AscribeUserTy` for use here?

Reply:
yes, I think so :+1: while I don't think is exploitable as we only use contravariance in `AscribeUserType` for places without any projections. Using the following in `super_ascribe_user_ty` does not cause any failures
```rust
                if variance != $(& $mutability)? ty::Variance::Covariant && !place.projection.is_empty() {
                    bug!()
                }
```
we should still add the variance to the use context though as this isn't something I want to implicitly rely on.